```shell
python "/share/compling/projects/sanpi/script/sample_pickle.py" \
    -p "${DATA_DIR}neg-bigrams_thr0-001p.35f.pkl.gz" \
    -f 'neg_deprel!=neg' -f 'neg_deprel!=advmod' -f 'neg_deprel!=nsubj' -f 'neg_deprel!=case' \
    -f 'neg_deprel!=cc' -f 'neg_deprel!=cc:preconj' -f 'neg_deprel!=cop' -f 'neg_deprel!=det' \
    -c 'neg_deprel' -c 'neg_form_lower' -c 'adv_form_lower' -c 'adj_form_lower' -c 'text_window' \
    -f 'neg_head==ADJ' -N 10000 -q -C > "${DATA_DIR}sample-35f_unk-deprel_adj-head.10k.csv"
```

In [1]:
import pandas as pd
from pathlib import Path
from utils import sample_pickle

DATA_DIR=Path("/share/compling/data/sanpi/4_post-processed/RBdirect")
COLUMNS=['neg_deprel', 'neg_form_lower', 'adv_form_lower', 'adj_form_lower', 'text_window']
HEAD = 'NEG'
N_FILES_PROC = 35
N_ROWS = 10000

#> Loaded variable 'df' from URI: /share/compling/data/sanpi/4_post-processed/RBdirect/sample-35f_unk-deprel_adj-head.1k.csv
# df = pd.read_csv(r'/share/compling/data/sanpi/4_post-processed/RBdirect/sample-35f_unk-deprel_adj-head.5k.csv')


## Define permitted `neg_deprel` values by head node

From [Oct 23 log](https://gist.github.com/AndreaTheLinguist/ef895a6e8602a9a82ff729f219bc5628#running-notes-monday-october-23-2023):

### `NEG` _head_

i.e., `NEG` ➡️ `ADJ`

> | permit ✅                                                  | prohibit ⛔                                                 |
> |-------------------------------------------------------------|------------------------------------------------------------|
> | `amod`, `ccomp`, `csubj`, `dep`, `infomod`, `pcomp`, `pobj` | `acl:relcl`, `advcl`, `appos`, `conj`, `nmod`, `parataxis` |



In [2]:
OK_DEPREL = {
    'ADJ': ['neg', 'advmod', 'nsubj', 'case', 'cop', 'det', 'cc', 'cc:preconj'], 
    'NEG': ['amod', 'ccomp', 'csubj', 'dep', 'infomod', 'pcomp', 'pobj']
    }

## Load data from composite cleaned and filtered table of hit info

### For 10 corpus parts and only 8 rows first, to illustrate usage of `sample_pickle`

Note that any sorting (as specified by `sort_by` parameter) is applie _after_ sample is drawn

In [3]:
def set_input_path(n_files: int = N_FILES_PROC) -> Path:
    return DATA_DIR.joinpath(f'neg-bigrams_thr0-001p.{n_files}f.pkl.gz')


def set_filters(head: str = HEAD,
                ok_deprel: dict = OK_DEPREL) -> list:

    # filters = [f'neg_deprel!=neg', 'neg_deprel!=advmod', 'neg_deprel!=nsubj', 'neg_deprel!=case',
    #            'neg_deprel!=cop', 'neg_deprel!=det', 'neg_deprel!=cc', 'neg_deprel!=cc:preconj'
    #            ]
    filters = [f'neg_deprel!={rel}' for rel in ok_deprel[head]]
    return filters + [f'neg_head=={head}']


def get_sample(data_path, filters, print_table, sample_size) -> pd.DataFrame:
    return sample_pickle(
        data_path=data_path,
        sample_size=sample_size,
        columns=COLUMNS,
        filters=filters,
        markdown=True,
        sort_by='neg_deprel',
        print_sample=print_table
    )


df = get_sample(data_path=set_input_path(n_files=10),
                filters=set_filters(),
                print_table=True,
                sample_size=8)


## Sampling from `/share/compling/data/sanpi/4_post-processed/RBdirect/neg-bigrams_thr0-001p.10f.pkl.gz`

- *filtering rows...*
  - ✓ Applied filter: `neg_deprel!=amod`
  - ✓ Applied filter: `neg_deprel!=ccomp`
  - ✓ Applied filter: `neg_deprel!=csubj`
  - ✓ Applied filter: `neg_deprel!=dep`
  - ✓ Applied filter: `neg_deprel!=infomod`
  - ✓ Applied filter: `neg_deprel!=pcomp`
  - ✓ Applied filter: `neg_deprel!=pobj`
  - ✓ Applied filter: `neg_head==NEG`

- *selecting columns...*

### 8 random rows matching filter(s) from `neg-bigrams_thr0-001p.10f.pkl.gz`

| hit_id                                | neg_deprel   | neg_form_lower   | adv_form_lower   | adj_form_lower   | text_window                                                                                                    |
|:--------------------------------------|:-------------|:-----------------|:-----------------|:-----------------|:-----------------------------------------------------------------------------------------------


## Sampling from `/share/compling/data/sanpi/4_post-processed/RBdirect/neg-bigrams_thr0-001p.10f.pkl.gz`

- *filtering rows...*
  - ✓ Applied filter: `neg_deprel!=neg`
  - ✓ Applied filter: `neg_deprel!=advmod`
  - ✓ Applied filter: `neg_deprel!=nsubj`
  - ✓ Applied filter: `neg_deprel!=case`
  - ✓ Applied filter: `neg_deprel!=cop`
  - ✓ Applied filter: `neg_deprel!=det`
  - ✓ Applied filter: `neg_deprel!=cc`
  - ✓ Applied filter: `neg_deprel!=cc:preconj`
  - ✓ Applied filter: `neg_head==ADJ`

- *selecting columns...*

### 8 random rows matching filter(s) from `neg-bigrams_thr0-001p.10f.pkl.gz`

| hit_id                                   | neg_deprel   | neg_form_lower   | adv_form_lower   | adj_form_lower   | text_window                                                                                                        |
|:-----------------------------------------|:-------------|:-----------------|:-----------------|:-----------------|:-------------------------------------------------------------------------------------------------------------------|
| apw_eng_20020208_0776_6:24-36-37         | dep          | without          | more             | difficult        | this course because , without them , the new equilibrium that we are seeking will be more difficult to attain , '' |
| nyt_eng_19990416_0358_28:08-11-12        | dep          | neither          | particularly     | complex          | n't easy , but neither was it particularly complex .                                                               |
| apw_eng_20090522_0774_13:2-3-4           | dep          | no               | more             | clear            | with no more clear chances for sprinters ,                                                                         |
| apw_eng_19970501_1204_15:24-25-26        | dep          | no               | longer           | afraid           | the Conservatives _ and no longer afraid of Labor .                                                                |
| pcc_eng_20_099.6848_x1594566_07:29-31-32 | mark         | without          | as               | blatant          | of the scenes ( without being as blatant as traditional porn )                                                     |
| pcc_eng_06_012.5761_x0187070_041:6-8-9   | mark         | without          | too              | sweet            | is rich and complex without being too sweet and comes in a                                                         |
| pcc_eng_19_050.8210_x0804180_82:24-26-27 | mark         | without          | consciously      | aware            | greater quantity , sometimes without being consciously aware that learning is taking                               |
| pcc_eng_18_046.7718_x0740811_07:25-27-28 | mark         | without          | too              | overwhelming     | the students to shore without being too overwhelming .                                                             |


In [4]:
def clean_data(df):
    if 'hit_id' in df.columns:
        hit_id = df.hit_id
        df = df.set_index('hit_id')
    else: 
        hit_id = df.index.to_series()

    df['corpus'] = hit_id.str.split('_eng_').str.get(0)
    df['match_ix'] = hit_id.str.rsplit(':',1).str.get(1)
    return df

df_clean = clean_data(df.copy())
df_clean.head()

neg_deprel neg_form_lower  \
hit_id                                                            
nyt_eng_19981021_0419_31:24-35-36          acomp        nothing   
pcc_eng_08_077.0493_x1231298_14:4-5-6     advmod          ain't   
nyt_eng_19960919_0547_44:11-13-14           attr           none   
nyt_eng_19970825_0226_14:17-20-21         infmod        nothing   
apw_eng_20010517_0898_12:04-09-10          rcmod        nothing   

                                      adv_form_lower adj_form_lower  \
hit_id                                                                
nyt_eng_19981021_0419_31:24-35-36             fairly       balanced   
pcc_eng_08_077.0493_x1231298_14:4-5-6            too            bad   
nyt_eng_19960919_0547_44:11-13-14               less     impressive   
nyt_eng_19970825_0226_14:17-20-21                too        excited   
apw_eng_20010517_0898_12:04-09-10               more      important   

                                                                             text_window  \
hit_id                                                                                     
nyt_eng_19981021_0419_31:24-35-36      that the news is nothing more than a snippet h...   
pcc_eng_08_077.0493_x1231298_14:4-5-6          And that hair ain't too bad either , if I   
nyt_eng_19960919_0547_44:11-13-14      a flame that is none the less impressive for i...   
nyt_eng_19970825_0226_14:17-20-21      through Student Advantage are nothing to get t...   
apw_eng_20010517_0898_12:04-09-10      `` There 's nothing that could have been more ...   

                                      corpus  match_ix  
hit_id                                                  
nyt_eng_19981021_0419_31:24-35-36        nyt  24-35-36  
pcc_eng_08_077.0493_x1231298_14:4-5-6    pcc     4-5-6  
nyt_eng_19960919_0547_44:11-13-14        nyt  11-13-14  
nyt_eng_19970825_0226_14:17-20-21        nyt  17-20-21  
apw_eng_20010517_0898_12:04-09-10        apw  04-09-10

### Now with larger sample and printing suppressed

In [5]:
df = get_sample(data_path=set_input_path(n_files=N_FILES_PROC),
                filters=set_filters(),
                print_table=False,
                sample_size=10000)
df_clean = clean_data(df.copy())


## Sampling from `/share/compling/data/sanpi/4_post-processed/RBdirect/neg-bigrams_thr0-001p.35f.pkl.gz`

- *filtering rows...*
  - ✓ Applied filter: `neg_deprel!=amod`
  - ✓ Applied filter: `neg_deprel!=ccomp`
  - ✓ Applied filter: `neg_deprel!=csubj`
  - ✓ Applied filter: `neg_deprel!=dep`
  - ✓ Applied filter: `neg_deprel!=infomod`
  - ✓ Applied filter: `neg_deprel!=pcomp`
  - ✓ Applied filter: `neg_deprel!=pobj`
  - ✓ Applied filter: `neg_head==NEG`

- *selecting columns...*

### All (1675) row(s) matching filter(s) from `neg-bigrams_thr0-001p.35f.pkl.gz`




## Sampling from `/share/compling/data/sanpi/4_post-processed/RBdirect/neg-bigrams_thr0-001p.35f.pkl.gz`

- *filtering rows...*
  - ✓ Applied filter: `neg_deprel!=amod`
  - ✓ Applied filter: `neg_deprel!=ccomp`
  - ✓ Applied filter: `neg_deprel!=csubj`
  - ✓ Applied filter: `neg_deprel!=dep`
  - ✓ Applied filter: `neg_deprel!=infomod`
  - ✓ Applied filter: `neg_deprel!=pcomp`
  - ✓ Applied filter: `neg_deprel!=pobj`
  - ✓ Applied filter: `neg_head==NEG`

- *selecting columns...*

### All (1675) row(s) matching filter(s) from `neg-bigrams_thr0-001p.35f.pkl.gz`



In [6]:
df_clean

neg_deprel neg_form_lower  \
hit_id                                                                 
pcc_eng_02_035.2935_x0555063_245:3-6-7            acl        nothing   
pcc_eng_27_108.00020_x1731542_138:13-21-22        acl        nothing   
pcc_eng_06_108.8040_x1743933_06:5-8-9             acl        nothing   
pcc_eng_02_097.3703_x1558033_033:05-13-14         acl        nothing   
pcc_eng_18_026.5996_x0413938_56:17-23-24          acl        nothing   
...                                               ...            ...   
pcc_eng_11_034.0959_x0535597_26:16-17-18        xcomp          ain't   
pcc_eng_18_073.9710_x1181785_06:17-18-19        xcomp          ain't   
pcc_eng_12_005.7442_x0076307_24:17-18-19        xcomp          ain't   
pcc_eng_16_015.1445_x0229183_2:13-14-15         xcomp          ain't   
pcc_eng_26_018.1104_x0276567_20:44-46-47        xcomp           aint   

                                           adv_form_lower adj_form_lower  \
hit_id                                                                     
pcc_eng_02_035.2935_x0555063_245:3-6-7           normally           busy   
pcc_eng_27_108.00020_x1731542_138:13-21-22     absolutely      confident   
pcc_eng_06_108.8040_x1743933_06:5-8-9              overly      concerned   
pcc_eng_02_097.3703_x1558033_033:05-13-14      absolutely  extraordinary   
pcc_eng_18_026.5996_x0413938_56:17-23-24              n't        capable   
...                                                   ...            ...   
pcc_eng_11_034.0959_x0535597_26:16-17-18              too         shabby   
pcc_eng_18_073.9710_x1181785_06:17-18-19             very          happy   
pcc_eng_12_005.7442_x0076307_24:17-18-19             that          scary   
pcc_eng_16_015.1445_x0229183_2:13-14-15                so            bad   
pcc_eng_26_018.1104_x0276567_20:44-46-47               as           good   

                                                                                  text_window  \
hit_id                                                                                          
pcc_eng_02_035.2935_x0555063_245:3-6-7          They have nothing to be normally busy about .   
pcc_eng_27_108.00020_x1731542_138:13-21-22  that there is truly nothing in this world that...   
pcc_eng_06_108.8040_x1743933_06:5-8-9       ' This slowdown is nothing to be overly concer...   
pcc_eng_02_097.3703_x1558033_033:05-13-14   1 . There is nothing excellent amongst men , w...   
pcc_eng_18_026.5996_x0413938_56:17-23-24    extremes , it portrays nothing that a human be...   
...                                                                                       ...   
pcc_eng_11_034.0959_x0535597_26:16-17-18    by all accounts she ain't too shabby in the ac...   
pcc_eng_18_073.9710_x1181785_06:17-18-19    fact the players still ain't very happy about ...   
pcc_eng_12_005.7442_x0076307_24:17-18-19       era , it simply ain't that scary anymore . . .   
pcc_eng_16_015.1445_x0229183_2:13-14-15               gets and her face ain't so bad either !   
pcc_eng_26_018.1104_x0276567_20:44-46-47    substitute for Pride but aint nearly as good ,...   

                                           corpus  match_ix  
hit_id                                                       
pcc_eng_02_035.2935_x0555063_245:3-6-7        pcc     3-6-7  
pcc_eng_27_108.00020_x1731542_138:13-21-22    pcc  13-21-22  
pcc_eng_06_108.8040_x1743933_06:5-8-9         pcc     5-8-9  
pcc_eng_02_097.3703_x1558033_033:05-13-14     pcc  05-13-14  
pcc_eng_18_026.5996_x0413938_56:17-23-24      pcc  17-23-24  
...                                           ...       ...  
pcc_eng_11_034.0959_x0535597_26:16-17-18      pcc  16-17-18  
pcc_eng_18_073.9710_x1181785_06:17-18-19      pcc  17-18-19  
pcc_eng_12_005.7442_x0076307_24:17-18-19      pcc  17-18-19  
pcc_eng_16_015.1445_x0229183_2:13-14-15       pcc  13-14-15  
pcc_eng_26_018.1104_x0276567_20:44-46-47      pcc  44-46-47  

[1675 rows x 7 columns]

In [13]:
deprel_counts=df_clean.value_counts('neg_deprel', normalize=True)
print(deprel_counts.multiply(100).round(2).to_frame('% of unknown relation NEG-head hits').to_markdown(floatfmt=',.2f'))

| neg_deprel   |   % of unknown relation NEG-head hits |
|:-------------|--------------------------------------:|
| xcomp        |                                 38.27 |
| acl          |                                 26.93 |
| rcmod        |                                 14.33 |
| obl          |                                  9.37 |
| advmod       |                                  3.64 |
| infmod       |                                  2.87 |
| obj          |                                  2.63 |
| acomp        |                                  0.72 |
| nsubj        |                                  0.72 |
| attr         |                                  0.24 |
| dobj         |                                  0.06 |
| list         |                                  0.06 |
| neg          |                                  0.06 |
| partmod      |                                  0.06 |
| prep         |                                  0.06 |


| neg_deprel   |   % of unknown relation NEG-head hits |
|:-------------|--------------------------------------:|
| xcomp        |                                 38.27 |
| acl          |                                 26.93 |
| rcmod        |                                 14.33 |
| obl          |                                  9.37 |
| advmod       |                                  3.64 |
| infmod       |                                  2.87 |
| obj          |                                  2.63 |
| acomp        |                                  0.72 |
| nsubj        |                                  0.72 |
| attr         |                                  0.24 |
| dobj         |                                  0.06 |
| list         |                                  0.06 |
| neg          |                                  0.06 |
| partmod      |                                  0.06 |
| prep         |                                  0.06 |


In [16]:
for g, d in df_clean[['neg_deprel', 'neg_form_lower', 'corpus']].groupby('neg_deprel'):
    vc = d.value_counts(normalize=True).multiply(100).round(2).to_frame(
        f'% of `{g}` hits')
    vc['count']=d.value_counts()
    vc = vc.reset_index().iloc[:, 1:].set_index('neg_form_lower')

    print(f'`neg_deprel == {g}`: _{len(d)} total in NEG-head hits w/ unknown relation ({round(len(d)/len(df_clean) * 100, 2)}%)_')
    print(vc.to_markdown(floatfmt=',.1f'))
    print('\n---\n')

`neg_deprel == acl`: _451 total in NEG-head hits w/ unknown relation (26.93%)_
| neg_form_lower   | corpus   |   % of `acl` hits |   count |
|:-----------------|:---------|------------------:|--------:|
| nothing          | pcc      |              90.0 |     406 |
| none             | pcc      |               6.7 |      30 |
| few              | pcc      |               2.0 |       9 |
| nobody           | pcc      |               0.9 |       4 |
| no               | pcc      |               0.4 |       2 |

---

`neg_deprel == acomp`: _12 total in NEG-head hits w/ unknown relation (0.72%)_
| neg_form_lower   | corpus   |   % of `acomp` hits |   count |
|:-----------------|:---------|--------------------:|--------:|
| nothing          | nyt      |                58.3 |       7 |
| aint             | nyt      |                16.7 |       2 |
| few              | apw      |                 8.3 |       1 |
| few              | nyt      |                 8.3 |       1 |
| nothing         

`neg_deprel == acl`: _451 total in NEG-head hits w/ unknown relation (26.93%)_
| neg_form_lower   | corpus   |   % of `acl` hits |   count |
|:-----------------|:---------|------------------:|--------:|
| nothing          | pcc      |              90.0 |     406 |
| none             | pcc      |               6.7 |      30 |
| few              | pcc      |               2.0 |       9 |
| nobody           | pcc      |               0.9 |       4 |
| no               | pcc      |               0.4 |       2 |

---

`neg_deprel == acomp`: _12 total in NEG-head hits w/ unknown relation (0.72%)_
| neg_form_lower   | corpus   |   % of `acomp` hits |   count |
|:-----------------|:---------|--------------------:|--------:|
| nothing          | nyt      |                58.3 |       7 |
| aint             | nyt      |                16.7 |       2 |
| few              | apw      |                 8.3 |       1 |
| few              | nyt      |                 8.3 |       1 |
| nothing          | apw      |                 8.3 |       1 |

---

`neg_deprel == advmod`: _61 total in NEG-head hits w/ unknown relation (3.64%)_
| neg_form_lower   | corpus   |   % of `advmod` hits |   count |
|:-----------------|:---------|---------------------:|--------:|
| ain't            | pcc      |                 34.4 |      21 |
| nothing          | nyt      |                 16.4 |      10 |
| none             | nyt      |                 11.5 |       7 |
| few              | pcc      |                  8.2 |       5 |
| aint             | pcc      |                  4.9 |       3 |
| few              | nyt      |                  4.9 |       3 |
| fewer            | pcc      |                  3.3 |       2 |
| nothing          | pcc      |                  3.3 |       2 |
| few              | apw      |                  1.6 |       1 |
| neither          | nyt      |                  1.6 |       1 |
| no               | pcc      |                  1.6 |       1 |
| nod              | pcc      |                  1.6 |       1 |
| none             | apw      |                  1.6 |       1 |
| not              | nyt      |                  1.6 |       1 |
| not              | pcc      |                  1.6 |       1 |
| nothing          | apw      |                  1.6 |       1 |

---

`neg_deprel == attr`: _4 total in NEG-head hits w/ unknown relation (0.24%)_
| neg_form_lower   | corpus   |   % of `attr` hits |   count |
|:-----------------|:---------|-------------------:|--------:|
| none             | nyt      |              100.0 |       4 |

---

`neg_deprel == dobj`: _1 total in NEG-head hits w/ unknown relation (0.06%)_
| neg_form_lower   | corpus   |   % of `dobj` hits |   count |
|:-----------------|:---------|-------------------:|--------:|
| few              | nyt      |              100.0 |       1 |

---

`neg_deprel == infmod`: _48 total in NEG-head hits w/ unknown relation (2.87%)_
| neg_form_lower   | corpus   |   % of `infmod` hits |   count |
|:-----------------|:---------|---------------------:|--------:|
| nothing          | nyt      |                 54.2 |      26 |
| nothing          | apw      |                 39.6 |      19 |
| few              | nyt      |                  4.2 |       2 |
| few              | apw      |                  2.1 |       1 |

---

`neg_deprel == list`: _1 total in NEG-head hits w/ unknown relation (0.06%)_
| neg_form_lower   | corpus   |   % of `list` hits |   count |
|:-----------------|:---------|-------------------:|--------:|
| none             | pcc      |              100.0 |       1 |

---

`neg_deprel == neg`: _1 total in NEG-head hits w/ unknown relation (0.06%)_
| neg_form_lower   | corpus   |   % of `neg` hits |   count |
|:-----------------|:---------|------------------:|--------:|
| not              | nyt      |             100.0 |       1 |

---

`neg_deprel == nsubj`: _12 total in NEG-head hits w/ unknown relation (0.72%)_
| neg_form_lower   | corpus   |   % of `nsubj` hits |   count |
|:-----------------|:---------|--------------------:|--------:|
| aint             | pcc      |                33.3 |       4 |
| ain't            | pcc      |                25.0 |       3 |
| not              | nyt      |                25.0 |       3 |
| few              | nyt      |                 8.3 |       1 |
| few              | pcc      |                 8.3 |       1 |

---

`neg_deprel == obj`: _44 total in NEG-head hits w/ unknown relation (2.63%)_
| neg_form_lower   | corpus   |   % of `obj` hits |   count |
|:-----------------|:---------|------------------:|--------:|
| ain't            | pcc      |              90.9 |      40 |
| aint             | pcc      |               9.1 |       4 |

---

`neg_deprel == obl`: _157 total in NEG-head hits w/ unknown relation (9.37%)_
| neg_form_lower   | corpus   |   % of `obl` hits |   count |
|:-----------------|:---------|------------------:|--------:|
| few              | pcc      |              89.2 |     140 |
| fewer            | pcc      |               5.1 |       8 |
| not              | pcc      |               2.5 |       4 |
| ain't            | pcc      |               1.3 |       2 |
| aint             | pcc      |               0.6 |       1 |
| n't              | pcc      |               0.6 |       1 |
| nothing          | pcc      |               0.6 |       1 |

---

`neg_deprel == partmod`: _1 total in NEG-head hits w/ unknown relation (0.06%)_
| neg_form_lower   | corpus   |   % of `partmod` hits |   count |
|:-----------------|:---------|----------------------:|--------:|
| nothing          | nyt      |                 100.0 |       1 |

---

`neg_deprel == prep`: _1 total in NEG-head hits w/ unknown relation (0.06%)_
| neg_form_lower   | corpus   |   % of `prep` hits |   count |
|:-----------------|:---------|-------------------:|--------:|
| few              | nyt      |              100.0 |       1 |

---

`neg_deprel == rcmod`: _240 total in NEG-head hits w/ unknown relation (14.33%)_
| neg_form_lower   | corpus   |   % of `rcmod` hits |   count |
|:-----------------|:---------|--------------------:|--------:|
| nothing          | nyt      |                36.7 |      88 |
| few              | nyt      |                22.1 |      53 |
| nothing          | apw      |                17.1 |      41 |
| none             | nyt      |                 7.5 |      18 |
| nobody           | nyt      |                 6.7 |      16 |
| few              | apw      |                 4.2 |      10 |
| none             | apw      |                 2.1 |       5 |
| no               | nyt      |                 1.7 |       4 |
| neither          | nyt      |                 0.8 |       2 |
| nobody           | apw      |                 0.8 |       2 |
| nots             | nyt      |                 0.4 |       1 |

---

`neg_deprel == xcomp`: _641 total in NEG-head hits w/ unknown relation (38.27%)_
| neg_form_lower   | corpus   |   % of `xcomp` hits |   count |
|:-----------------|:---------|--------------------:|--------:|
| ain't            | pcc      |                84.7 |     543 |
| aint             | pcc      |                 8.1 |      52 |
| nit              | pcc      |                 2.6 |      17 |
| not              | pcc      |                 1.2 |       8 |
| nothing          | pcc      |                 0.8 |       5 |
| without          | pcc      |                 0.8 |       5 |
| few              | apw      |                 0.3 |       2 |
| no               | pcc      |                 0.3 |       2 |
| few              | nyt      |                 0.2 |       1 |
| few              | pcc      |                 0.2 |       1 |
| nobody           | pcc      |                 0.2 |       1 |
| nod              | pcc      |                 0.2 |       1 |
| nothing          | apw      |                 0.2 |       1 |
| nothing          | nyt      |                 0.2 |       1 |
| nothingis        | pcc      |                 0.2 |       1 |


In [18]:
for deprel in deprel_counts[deprel_counts>0.01].index:
    
    # print(f'\n- **P** `{deprel}`')
    print(f'\n#### Sample of `{deprel}` relations (with `ADJ` head)\n')
    selection = df_clean[df_clean.neg_deprel==deprel]
    print(selection.sample(min(len(selection),25)).filter(regex=r'(lower|window|ix)').set_index('match_ix').to_markdown())


#### Sample of `xcomp` relations (with `ADJ` head)

| match_ix   | neg_form_lower   | adv_form_lower   | adj_form_lower   | text_window                                                            |
|:-----------|:-----------------|:-----------------|:-----------------|:-----------------------------------------------------------------------|
| 26-27-28   | ain't            | no               | better           | Trade Union , he ain't no better than a milestone for                  |
| 3-4-5      | ain't            | as               | bad              | It really ain't as bad as it seems .                                   |
| 5-6-7      | ain't            | too              | tight            | Make sure your skirt ain't too tight                                   |
| 6-7-8      | ain't            | too              | shabby           | $ 49 price tag ain't too shabby , either .                             |
| 25-26-27   | ain't            | all              | rich             | on re


#### Sample of `xcomp` relations (with `ADJ` head)

| match_ix   | neg_form_lower   | adv_form_lower   | adj_form_lower   | text_window                                                            |
|:-----------|:-----------------|:-----------------|:-----------------|:-----------------------------------------------------------------------|
| 26-27-28   | ain't            | no               | better           | Trade Union , he ain't no better than a milestone for                  |
| 3-4-5      | ain't            | as               | bad              | It really ain't as bad as it seems .                                   |
| 5-6-7      | ain't            | too              | tight            | Make sure your skirt ain't too tight                                   |
| 6-7-8      | ain't            | too              | shabby           | $ 49 price tag ain't too shabby , either .                             |
| 25-26-27   | ain't            | all              | rich             | on reason bein we ain't all rich And I 'ma be                          |
| 6-7-8      | ain't            | so               | big              | Jesse Branson , you ain't so big I wo n't put                          |
| 2-3-4      | aint             | as               | bad              | You aint as bad as me I get                                            |
| 29-30-31   | ain't            | as               | black            | is that black holes ain't as black as they are painted                 |
| 7-8-9      | ain't            | as               | strong           | wind in these parts ain't as strong as you might think                 |
| 20-21-22   | ain't            | as               | cool             | cool , but he ain't as cool as his alter- ego                          |
| 6-7-8      | ain't            | too              | bad              | rest of the cast ain't too bad either .                                |
| 6-7-8      | ain't            | that             | bad              | , frankly , still ain't that bad for what are 30                       |
| 7-8-9      | ain't            | too              | bad              | rest of my team ain't too bad either , but they                        |
| 7-8-9      | aint             | that             | bad              | Drake , he really aint that bad .                                      |
| 17-18-19   | ain't            | so               | bad              | that this life really ain't so bad .                                   |
| 14-15-16   | ain't            | half             | bad              | means that my book ain't half bad ?                                    |
| 22-23-24   | ain't            | as               | good             | somewhat entertaining , they ain't as good as Dune and the             |
| 3-5-6      | ain't            | so               | bright           | But they ain't half so bright                                          |
| 2-6-7      | ain't            | more             | everyday         | Things ain't the same no more Everyday my city seems to                |
| 34-35-36   | ain't            | so               | noble            | the noble lawyer who ain't so noble after all .                        |
| 08-09-10   | ain't            | as               | good             | have just enough I ain't as good as I once was                         |
| 27-28-29   | ain't            | that             | bad              | be told , they ain't that bad .                                        |
| 5-6-7      | ain't            | too              | bad              | Heck , the buses ain't too bad either .                                |
| 15-16-17   | ain't            | that             | easy             | Hunt , but it ain't that easy .                                        |
| 5-7-8      | nit              | that             | bad              | The non-conference schedule is nit all that bad after starting out the |

#### Sample of `acl` relations (with `ADJ` head)

| match_ix   | neg_form_lower   | adv_form_lower   | adj_form_lower   | text_window                                                                               |
|:-----------|:-----------------|:-----------------|:-----------------|:------------------------------------------------------------------------------------------|
| 24-30-31   | nothing          | too              | short            | ( all clean , nothing negative except that it is too short - 3yrs ) I                     |
| 15-23-24   | nothing          | already          | familiar         | . demonstrate there 's nothing really new there that we are n't already familiar with . " |
| 22-24-25   | nothing          | not              | lucid            | Euripides ' poetry is nothing if not lucid -- the petulant young                          |
| 19-23-24   | nothing          | utterly          | depressing       | useful -- to say nothing of its being utterly depressing -- survey of the                 |
| 13-15-16   | nothing          | not              | eccentric        | press , Revolver 's nothing if not eccentric ; at times ,                                 |
| 1-8-9      | none             | soooo            | badass           | None of that " John Wick is SOOOO BADASS , he wears various                               |
| 22-24-25   | nothing          | not              | cynical          | -- " I 'm nothing if not cynical and jaded " --                                           |
| 22-27-28   | nothing          | actually         | true             | internally coherent tells us nothing about whether it is actually true -- that is ,       |
| 3-6-7      | nothing          | really           | successful       | There 's nothing like being really successful and finally making lots                     |
| 26-32-33   | nothing          | quite            | open             | a normal friendship , nothing more than that i'd been quite open with him that i          |
| 03-10-11   | none             | too              | young            | There is none whole save such as are yet too young to read the Scripture                  |
| 07-14-15   | nothing          | too              | stupid           | drug test failure proves nothing other than that the guy is too stupid to even stick to   |
| 19-21-22   | nothing          | less             | risky            | , and see saying nothing as less risky than saying something ?                            |
| 03-09-10   | none             | too              | concerned        | It is none of our business to be too concerned                                            |
| 17-19-20   | nothing          | not              | consistent       | Father director has been nothing if not consistent , filing a divorce                     |
| 14-18-19   | nothing          | n't              | willing          | 's GOP there is nothing that they are n't willing to turn into a                          |
| 17-20-21   | nothing          | especially       | afraid           | Tulsa running game was nothing to be especially afraid of , making the                    |
| 1-3-4      | nobody           | however          | successful       | Nobody , however successful his past record ,                                             |
| 09-12-13   | nothing          | terribly         | concerned        | and there , but nothing to be terribly concerned with .                                   |
| 3-6-7      | nothing          | intrinsically    | proud            | There is nothing to be intrinsically proud of in bearing the                              |
| 23-30-31   | nothing          | very             | stressful        | me on board and nothing but the surrounding bush to be very stressful - he imagined large |
| 46-51-52   | nothing          | suddenly         | shocked          | , and often just nothing , when I was suddenly shocked to come across an                  |
| 26-32-33   | none             | most             | powerful         | but useless , with none being more so than the most powerful of them all ,                |
| 15-18-19   | none             | stupendously     | shocking         | Street , " but none quite as stupendously shocking as a scene we                          |
| 4-8-9      | nothing          | sexually         | energetic        | There is absolutely nothing incorrect with being sexually energetic however if you have   |

#### Sample of `rcmod` relations (with `ADJ` head)

| match_ix   | neg_form_lower   | adv_form_lower   | adj_form_lower   | text_window                                                                                                    |
|:-----------|:-----------------|:-----------------|:-----------------|:---------------------------------------------------------------------------------------------------------------|
| 15-18-19   | none             | statistically    | significant      | cholesterol readings but found none that were statistically significant .                                      |
| 4-7-8      | nothing          | ever             | frightened       | `` There was nothing he was ever frightened of , he was                                                        |
| 21-24-25   | few              | so               | underground      | line , even a few that are so underground that our own residents                                               |
| 05-13-14   | nothing          | broadly          | encouraging      | `` There was certainly nothing in what happened last night that is broadly encouraging . ''                    |
| 14-17-18   | nothing          | too              | worried          | but particularly there was nothing I was too worried about , '' Matsuzaka                                      |
| 11-21-22   | nothing          | any              | different        | China , there is nothing that Microsoft and Yahoo have to do that is any different from what Chinese companies |
| 08-12-13   | nothing          | that             | exciting         | of times it 's nothing that 's all that exciting . ''                                                          |
| 4-8-9      | nothing          | too              | happy            | so it 's nothing we can be too happy about . ''                                                                |
| 4-8-9      | nothing          | seriously        | radical          | `` There 's nothing whatever that 's seriously radical or progressive about hip-hop                            |
| 31-38-39   | nothing          | more             | emphatic         | High Frequency Economics , nothing that the Fed chairman had been more emphatic than at any time               |
| 08-12-13   | nothing          | more             | terrible         | that , there is nothing that could be more terrible to a system of                                             |
| 55-58-59   | few              | wildly           | successful       | lower prices of the few that become wildly successful .                                                        |
| 18-21-22   | nothing          | readily          | apparent         | edge in registration ? Nothing that is readily apparent .                                                      |
| 4-7-8      | nothing          | more             | onerous          | `` This is nothing that is more onerous than has been agreed                                                   |
| 11-14-15   | few              | really           | excellent        | there are still those few that are really excellent .                                                          |
| 4-7-8      | nothing          | really           | worried          | so it 's nothing we 're really worried about .                                                                 |
| 09-16-17   | nothing          | particularly     | concerned        | hung bridges , again nothing on our balance sheet we are particularly concerned about , '' he                  |
| 4-8-9      | nothing          | more             | concerned        | `` There is nothing that I 'm more concerned about than the safety                                             |
| 49-59-60   | few              | as               | great            | to name only a few whose impact on religious thought has been at least as great as that of the                 |
| 4-7-8      | nothing          | especially       | human            | `` There 's nothing that 's especially human about our new Windows                                             |
| 28-32-33   | few              | basically        | sedentary        | moderate exercisers and a few who had been basically sedentary . ''                                            |
| 20-24-25   | few              | wildly           | partisan         | just to name a few , who are wildly partisan and loud .                                                        |
| 06-15-16   | none             | as               | good             | me , ai n't none of the pop stuff happening that 's half as good as what he can                                |
| 4-8-9      | nothing          | any              | different        | so there 's nothing that 's really any different for him , or                                                  |
| 05-09-10   | nothing          | more             | destructive      | i can think of nothing that would be more destructive of what remaining faith                                  |

#### Sample of `obl` relations (with `ADJ` head)

| match_ix   | neg_form_lower   | adv_form_lower   | adj_form_lower   | text_window                                                                     |
|:-----------|:-----------------|:-----------------|:-----------------|:--------------------------------------------------------------------------------|
| 26-30-31   | few              | as               | many             | 103 , from as few as two to as many as one hundred .                            |
| 18-20-21   | few              | so               | many             | suffering inflicted by so few on so many that shook me up                       |
| 22-25-26   | few              | most             | popular          | Slash to name a few of the most popular - while they all                        |
| 07-10-11   | few              | most             | popular          | be focusing on a few of the most popular , to help familiarize                  |
| 09-12-13   | few              | most             | interesting      | 'll be covering a few of the most interesting of those discussions .            |
| 16-18-19   | few              | so               | little           | was done by so few with so little .                                             |
| 3-6-7      | few              | most             | important        | Choose a few of the most important and work them into                           |
| 14-16-17   | few              | elegantly        | incisive         | the victims , but few as elegantly incisive as Into The Abyss                   |
| 08-11-12   | few              | most             | popular          | above are just a few of the most popular and most well known                    |
| 17-20-21   | few              | more             | imperfect        | because he created a few of the more imperfect , but because of                 |
| 11-13-14   | few              | so               | many             | wrong done by so few to so many ?                                               |
| 23-27-28   | fewer            | still            | fewer            | women at undergraduate to fewer at postgraduate to still fewer at staff - while |
| 2-5-6      | few              | extremely        | few              | Very few among the extremely few I 've met actually                             |
| 20-23-24   | few              | politically      | sophisticated    | The Apprentice , very few except the politically sophisticated would be able to |
| 29-31-32   | n't              | not              | least            | bot and what is n't , not least because poor grammar and                        |
| 08-11-12   | few              | most             | interesting      | we pick just a few of the most interesting , the most notable                   |
| 15-18-19   | fewer            | less             | advantaged       | their resources , which fewer of the less advantaged could do .                 |
| 16-20-21   | few              | as               | many             | Britain range from as few as 20,000 to as many as 50,000 .                      |
| 15-18-19   | few              | most             | notable          | but check out a few of the most notable Are You Ready 's                        |
| 10-13-14   | few              | less             | gentle           | scene after hustling a few of the less gentle of gentlemen gamblers ,           |
| 08-11-12   | few              | most             | prevalent        | to mention only a few of the most prevalent , include increased profitability   |
| 11-14-15   | few              | most             | beneficial       | are capable of a few of the most beneficial out there service and               |
| 3-7-8      | not              | very             | least            | Were they not , at the very least , there would not                             |
| 27-30-31   | few              | very             | many             | ( just to mention few among the very many ) , are being                         |
| 3-6-7      | few              | critically       | essential        | These are few , by critically essential if we are to                            |

#### Sample of `advmod` relations (with `ADJ` head)

| match_ix    | neg_form_lower   | adv_form_lower   | adj_form_lower   | text_window                                                                              |
|:------------|:-----------------|:-----------------|:-----------------|:-----------------------------------------------------------------------------------------|
| 2-4-5       | few              | no               | more             | Very few - no more than about one -                                                      |
| 43-45-46    | aint             | any              | more             | more Jem , we aint not any more to meet ever like                                        |
| 05-10-11    | ain't            | no               | more             | Well , actually there ain't much of that , no more than usual .                          |
| 59-60-61    | ain't            | as               | bad              | Being a zombie really ain't as bad as you think ,                                        |
| 2-5-6       | ain't            | no               | more             | She Ain't a Child No More                                                                |
| 23-26-27    | ain't            | no               | more             | ain't with me he ain't with me no more                                                   |
| 23-25-26    | ain't            | no               | more             | the forest Cos it ain't there no more                                                    |
| 24-28-29    | few              | too              | few              | the escalators were too few for me , too few !                                           |
| 21-22-23    | nothing          | as               | literary         | such work , but nothing as literary .                                                    |
| 25-27-28    | nothing          | very             | little           | work force , making nothing or very little , while the top                               |
| 136-138-139 | ain't            | no               | more             | Up My Mind I Ain't Cheatin No More Because A Cheater Can't                               |
| 1-5-6       | nothing          | no               | more             | Nothing actually by dhimmi no more                                                       |
| 09-10-11    | nothing          | as               | restrictive      | latest drafts were `` nothing as restrictive as we thought it                            |
| 37-38-39    | few              | too              | many             | away ( maybe a few too many                                                              |
| 21-24-25    | ain't            | no               | more             | tell you , that ain't the case no more : thanks to the                                   |
| 09-10-11    | ain't            | as               | bad              | glasses off and he ain't as bad as people are making                                     |
| 1-2-3       | nothing          | much             | good             | nothing much good has happened since Boston                                              |
| 07-14-15    | nothing          | no               | worse            | BREATH OF WINTER IS NOTHING TO THEM THE FLAMETIDES OF SUMMER NO WORSE THEY FEAST AS THEY |
| 5-7-8       | ain't            | no               | more             | We both know we ain't kids no more                                                       |
| 13-14-15    | none             | as               | exciting         | , yes , but none as exciting , '' says Tina                                              |
| 11-14-15    | ain't            | no               | more             | ain't with me He ain't with me no more                                                   |
| 09-10-11    | ain't            | as               | bad              | away certain that things ain't as bad as I feared to                                     |
| 37-40-41    | ain't            | no               | more             | It Back , She Ain't a Child No More , Nobody 's Baby                                     |
| 1-2-3       | nothing          | much             | new              | nothing much new happens , and nothing                                                   |
| 25-26-27    | ain't            | as               | bad              | will suddenly realize he ain't as bad as his demeanor currently                          |

#### Sample of `infmod` relations (with `ADJ` head)

| match_ix   | neg_form_lower   | adv_form_lower   | adj_form_lower   | text_window                                                                |
|:-----------|:-----------------|:-----------------|:-----------------|:---------------------------------------------------------------------------|
| 3-6-7      | nothing          | all              | bent             | this is nothing to be all bent out of shape about                          |
| 33-36-37   | few              | consistently     | profitable       | is one of the few to be consistently profitable , with ambitions to        |
| 30-34-35   | nothing          | very             | excited          | the conference would be nothing to get `` very excited about '' because of |
| 13-16-17   | nothing          | as               | exciting         | retire there would be nothing to do as exciting or as challenging in       |
| 3-6-7      | nothing          | all              | uptight          | there 's nothing to get all uptight about .                                |
| 07-10-11   | nothing          | overly           | excited          | nothing to embrace , nothing to get overly excited about .                 |
| 11-14-15   | nothing          | more             | proud            | have nothing better , nothing to be more proud of , than German            |
| 08-11-12   | nothing          | too              | hyped            | me , there 's nothing to get too hyped about .                             |
| 24-27-28   | nothing          | too              | alarmed          | the blood work is nothing to get too alarmed about .                       |
| 04-09-10   | nothing          | very             | encouraged       | `` There 's nothing in here to be very encouraged about '' for the         |
| 11-14-15   | nothing          | super            | intense          | day , it 's nothing to get super intense about or worried over             |
| 17-20-21   | nothing          | too              | excited          | percent decrease overall `` nothing to get too excited about '' because of |
| 26-29-30   | few              | commercially     | viable           | no diamonds or too few to be commercially viable .                         |
| 4-8-9      | nothing          | so               | joyful           | '' He 's nothing really to be so joyful about . ''                         |
| 15-18-19   | nothing          | too              | concerned        | the report was `` nothing to get too concerned about , but the             |
| 20-23-24   | nothing          | too              | concerned        | a hairline crack , nothing to be too concerned about .                     |
| 08-11-12   | nothing          | really           | worried          | the June weakness is nothing to get really worried about .                 |
| 3-7-8      | nothing          | overly           | concerned        | there 's nothing to really be overly concerned about .                     |
| 12-15-16   | nothing          | majorly          | concerned        | and bruises , but nothing to be majorly concerned about . ''               |
| 25-28-29   | nothing          | so               | excited          | sexual activities were probably nothing to get so excited about .          |
| 04-09-10   | nothing          | too              | worried          | `` It 's nothing for me to get too worried about right now .               |
| 3-6-7      | nothing          | too              | worried          | there is nothing to be too worried about yet , ''                          |
| 28-31-32   | nothing          | really           | worried          | just a game , nothing to get really worried about .                        |
| 28-32-33   | nothing          | very             | excited          | the conference would be nothing to get `` very excited about . ''          |
| 17-20-21   | nothing          | too              | excited          | through Student Advantage are nothing to get too excited about .           |

#### Sample of `obj` relations (with `ADJ` head)

| match_ix   | neg_form_lower   | adv_form_lower   | adj_form_lower   | text_window                                             |
|:-----------|:-----------------|:-----------------|:-----------------|:--------------------------------------------------------|
| 27-28-29   | ain't            | much             | left             | for me But there ain't much left of me to see           |
| 06-09-10   | ain't            | no               | more             | friends tell me I ain't the same no more                |
| 10-11-12   | ain't            | so               | much             | is , that it ain't so much that a person goes           |
| 20-21-22   | ain't            | too              | much             | some pointers if it ain't too much of a hassle for      |
| 13-14-15   | ain't            | as               | fake             | be , and it ain't as fake as it could be                |
| 3-4-5      | ain't            | much             | more             | But it ain't much more we can do .                      |
| 5-6-7      | ain't            | so               | much             | " Aw , it ain't so much that he 's against              |
| 2-4-5      | ain't            | too              | much             | It ain't , too much to ask / but                        |
| 3-4-5      | ain't            | much             | more             | Maybe you ain't much more than a scenery ...            |
| 25-26-27   | ain't            | too              | many             | , When they 're ain't too many still doin ' this        |
| 2-3-4      | ain't            | much             | good             | There ain't much good on the other side                 |
| 3-4-5      | ain't            | as               | fun              | It just ain't as fun as the Clintons ,                  |
| 09-10-11   | ain't            | much             | more             | know , there really ain't much more I can say about     |
| 2-3-4      | aint             | much             | more             | There aint much more to it than that                    |
| 23-24-25   | ain't            | nearly           | enough           | very appealing , she ain't nearly enough .              |
| 4-5-6      | ain't            | exactly          | rich             | An ' they ain't exactly rich ' round these parts        |
| 7-8-9      | ain't            | really           | much             | in my hood there ain't really much to eat .             |
| 2-3-4      | ain't            | no               | bigger           | You ain't no bigger than a pint of                      |
| 23-24-25   | ain't            | hardly           | enough           | so far -- which ain't hardly enough ( though it 's      |
| 6-7-8      | aint             | too              | much             | you may think it aint too much .                        |
| 4-6-7      | ain't            | most             | worthy           | And while glass ain't the most worthy of elements for a |
| 10-11-12   | ain't            | over             | dramatic         | a whip and he ain't over dramatic and he is a           |
| 4-5-6      | ain't            | too              | much             | You know it ain't too much of a leap Now                |
| 2-3-4      | ain't            | much             | more             | There ain't much more for a married couple              |
| 3-4-5      | ain't            | too              | many             | And it ain't too many on it , I                         |
